### PR TITLE
rust 1.72.0

### PR DESCRIPTION
**Pull request by ZhongRuoyu**
Also, implement the following based on discussion in Slack:
- Avoid double `cargo` build and installation by replacing the copy in the original source.
- Use brewed `llvm`, `libgit2`, and `libssh2`.
- Explicitly specify the extra tools to build. This allows us to drop bundled `rust-analyzer` and `rustfmt` and revive the respective formulae.
- Various cleanups.

---

Failures: https://github.com/Homebrew/homebrew-core/pull/140367#issuecomment-1699281554

### CODE REVIEW

File: Formula/r/rust.rb [hunk 1]
```diff
@@ -98,21 +120,6 @@ def install
...
 
-    resource("cargobootstrap").stage do
-      system "./install.sh", "--prefix=#{buildpath}/cargobootstrap"
-    end
-    ENV.prepend_path "PATH", buildpath/"cargobootstrap/bin"
-
-    resource("cargo").stage do
-      ENV["RUSTC"] = bin/"rustc"
-      args = %W[--root #{prefix} --path . --force]
-      args += %w[--features curl-sys/force-system-lib-on-osx] if OS.mac?
```
Comment (ZhongRuoyu) [inline Formula/r/rust.rb:109]: We've lost system `curl` linkage. Need to figure out how to re-enable it.

Comment (ZhongRuoyu) [reply Formula/r/rust.rb:109]: Did an `inreplace`.

File: Formula/r/rust.rb [hunk 2]
```diff
@@ -36,6 +36,9 @@ class Rust < Formula
   depends_on "cmake" => :build
   depends_on "ninja" => :build
   depends_on "python@3.11" => :build
+  depends_on "libgit2"
+  depends_on "libssh2"
+  depends_on "llvm"
```
Comment (cho-m) [inline Formula/r/rust.rb:41]: Dependency loop on Linux: `llvm -> swig (build) -> ruby (test) -> rust (build)`

Comment (ZhongRuoyu) [reply Formula/r/rust.rb:41]: Can probably use `python` instead of `ruby` to test `swig`. Will look into that.

Comment (ZhongRuoyu) [reply Formula/r/rust.rb:41]: #140437

**Comment by ZhongRuoyu** [comment]
Cancelled CI as it took unusually long. Turned out that all runners were stuck at `brew install --build-from-source sh4d0wup`. 🤔

**Comment by chenrui333** [comment]
update cargo to [0.73.1](https://github.com/rust-lang/cargo/commit/26bba48309d080ef3a3a00053f4e1dc879ef1de9)

**Comment by chenrui333** [comment]
gonna check out sh4d0wup to see if it builds out fine before next rerun.

File: Formula/r/rust.rb [hunk 3]
```diff
@@ -4,14 +4,24 @@ class Rust < Formula
...
+
+    # Fix for a compiler issue that may cause some builds to spin forever.
+    # Remove on 1.73.0 release.
+    # https://github.com/rust-lang/rust/pull/114948
+    # https://github.com/rust-lang/rust/issues/115297
+    # https://github.com/kpcyrd/sh4d0wup/issues/12
+    patch do
+      url "https://github.com/rust-lang/rust/commit/0f7f6b70617fbcda9f73755fa9b560bfb0a588eb.patch?full_index=1"
+      sha256 "549f446612bef10a1a06a967f61655b59c1b6895d762d764ca89caf65c114fe9"
     end
```
Comment (ZhongRuoyu) [inline Formula/r/rust.rb:25]: Confirmed that this fixed the `sh4d0wup` issue. We should be good for another CI run.

**Comment by ZhongRuoyu** [comment]
## Source build failures

- [ ] `clarinet` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:2798)): `error[E0080]: evaluation of constant value failed`
- [x] `cloudflare-quiche` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:2988)): ``error: no matching package named `boring` found``; #140170
- [ ] `flux` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:4286)): `error: private item shadows public glob re-export`
- [ ] `gluon` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:4996)): `error[E0308]: mismatched types`
- [ ] `grin-wallet` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:5376)): `error: variable does not need to be mutable`
- [ ] `influxdb` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:5937)): `error: private item shadows public glob re-export`
- [ ] `influxdb@1` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:6191)): `error: private item shadows public glob re-export`
- [ ] `kapacitor` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:6635)): `error: private item shadows public glob re-export`
- [ ] `leaf-proxy` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:7017)): `error[E0308]: mismatched types`
- [ ] (x86_64) `pushpin` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950652?pr=140367#step:3:8405)): `fatal error: 'assert.h' file not found`
- [ ] `smartdns` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:9418)): ``error[E0277]: the trait bound `connection_provider::TokioRuntimeProvider: ConnectionProvider` is not satisfied in `dns_client::NameServer` ``
- [ ] (x86_64) `sn0int` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950652?pr=140367#step:3:10417)): ``error[E0275]: overflow evaluating the requirement `[closure@/Users/brew/Library/Caches/Homebrew/cargo_cache/registry/src/index.crates.io-6f17d22bba15001f/geo-0.23.1/src/algorithm/map_coords.rs:855:69: 855:72]: Fn<(geo_types::Coord<T>,)>` ``
- [ ] `tremor-runtime` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:10606)): ``error: no rules expected the token `:` ``

## Pre-existing issues

- [ ] `azure-cli` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:1649)): `error in setup.cfg: command 'bdist_wheel' has no such option 'azure_namespace_package'`; Azure/azure-cli#27047
- [ ] `roapi` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:8697)): ``error: failed to get `connectorx` as a dependency of package `columnq v0.5.0 (/private/tmp/roapi-20230828-80688-uw1a7h/roapi-roapi-v0.9.0/columnq)` ``; roapi/roapi#287
- [ ] `rustfmt` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:9000)): `rustfmt is not linked`
- [ ] `sslyze` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:9964)): `urllib.error.HTTPError: HTTP Error 404: Not Found`; #138916
- [ ] (x86_64) `watchman` ([logs](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950652?pr=140367#step:3:12364)): `Undefined symbols for architecture x86_64`; https://github.com/Homebrew/homebrew-core/pull/118242#issuecomment-1367052317

## Transient failures

- `condure` ([12-x86_64](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950798?pr=140367#step:3:3277), [11-x86_64](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950957?pr=140367#step:3:3274)): `Timeout::Error: execution expired`
- `fnm` ([12-x86_64](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950798?pr=140367#step:3:4917)): `error: Can't download the requested binary: error sending request for url (https://nodejs.org/dist/v19.0.1/node-v19.0.1-darwin-x64.tar.xz): operation timed out`
- `gstreamer` ([13-arm64](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950726?pr=140367#step:3:5535), [11-arm64](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259951042?pr=140367#step:3:5889)): 504 error from <https://gitlab.freedesktop.org/gstreamer/gstreamer-rs>
- `pueue` ([11-arm64](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259951042?pr=140367#step:3:8635)): 500 error from <https://static.crates.io/crates/rev_buf_reader/rev_buf_reader-0.3.0.crate>
- `rtx` ([13-x86_64](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950652?pr=140367#step:3:9289)): `curl: (92) HTTP/2 stream 0 was not closed cleanly: INTERNAL_ERROR (err 2)`
- `volta` ([11-x86_64](https://github.com/Homebrew/homebrew-core/actions/runs/5994415857/job/16259950957?pr=140367#step:3:12648)): `Timeout::Error: execution expired`

**Comment by ZhongRuoyu** [comment]
The failures above are non-blocking as they don't affect the existing bottles. I've pushed a final update to finish this off (cleanups + unbundling `rust-analyzer` and `rustfmt`), and we'll be good to 🚢 this.

File: Formula/r/rust.rb [hunk 4]
```diff
@@ -45,28 +58,27 @@ class Rust < Formula
   conflicts_with "rust-analyzer", because: "both install `rust-analyzer` binary"
   conflicts_with "rustfmt", because: "both install `cargo-fmt` and `rustfmt` binaries"
```
Comment (ZhongRuoyu) [inline Formula/r/rust.rb:59]: Will follow up with a PR that removes these conflicts.

**Review by chenrui333 — APPROVED** [review]

File: Formula/r/rust.rb (hunk 4 above)
Comment (ZhongRuoyu) [reply Formula/r/rust.rb:59]: #140980

File: Formula/r/rust.rb [hunk 5]
```diff
@@ -4,25 +4,35 @@ class Rust < Formula
...
-    sha256 cellar: :any,                 big_sur:        "e204be3861c7f7df3f9ae0d56ffb1e41f0c865157784eb25933d2510ad0fa5d1"
-    sha256 cellar: :any_skip_relocation, x86_64_linux:   "6ab9ba5c6d96bf35e29d7d95756f02c0e29a661e3907b4d08a0e85a6996b4686"
+    sha256                               arm64_ventura:  "28e026c3d24f6dce2c1fb5087cf35cc5ffd88aa9055de215b81ae33d17e1f33b"
+    sha256                               arm64_monterey: "9a1ee1b72f0e6c6a0c1aa4d44de57ad18616710233c18e1e24960b9219d448aa"
+    sha256                               arm64_big_sur:  "7a65f06827c04b4d5ec27ff88efd143bc970081225eede6db71d2ce888ddf6ef"
+    sha256                               ventura:        "dbc20ef5f433182b31d7b2bf0f2bb29c96abcf2ecf2cccfc00c9774b46c27038"
+    sha256                               monterey:       "2a0ec660efed84e2ccfaaed4c30d2fc0577cbb9a41df32b374d8706202793837"
+    sha256                               big_sur:        "daaa3f24643f928e4098dae014c9e2055057a05f4c12710b837991808c59b193"
+    sha256 cellar: :any_skip_relocation, x86_64_linux:   "1701167773a55ba430ef5c318cb951f38041387949aed7f37a43b909e78a607c"
   end
```
Comment (ZhongRuoyu) [inline Formula/r/rust.rb:36]: As a side effect of using brewed `llvm`, we've lost relocatable bottles. 🥲

https://github.com/rust-lang/rust/blob/1.72.0/compiler/rustc_codegen_llvm/src/context.rs#L169-L196

**Pull request merged**